### ISSUE TITLE
It only uses one GPU for training

### ISSUE THREAD

**Issue by miladf2** [balancap/SSD-Tensorflow#9]
I am running the train_ssd_network.py code as it is and it only uses one GPU from the 8 GPUs that I have. Can u tell me how to scale it?

**Comment by miladf2** [comment]
I fixed this by changing one line in train_ssd.py   and adding soft replacement:        config = tf.ConfigProto(log_device_placement=False,allow_soft_placement=True, gpu_options=gpu_options)

**Comment by miladf2** [comment]
Still not sure if i am getting any benefit from scaling this to 8 GPUs though.. I had to reduce the batch sizes when I increased the number of GPUs to 8.

**Comment by balancap** [comment]
Hello,
Unfortunately, I do not have access to a multi-GPUs configuration for testing out myself.

The optimizer should be written in a way to split the load between the GPUs and average the gradients at every step, so it seems weird to have to reduce the batch sizes.
I basically took the imagenet model of TF slim https://github.com/tensorflow/models/tree/master/slim and adapted it to the SSD networks. 
I may have change something in the training script  which makes it buggy now :/

What you may try is to remove all the GPUs options and set `log_device_placement=True`. If you can upload the output, I can checked if the model is well-deployed on all GPUs.

**Comment by miladf2** [comment]
I tired may different options. with some combinations it uses part of the other gpus but doesnt use fully utilized them meaning gpu memory usage stays below 550MB. It sometimes breaks too.  I will copy the log.

**Comment by balancap** [comment]
That sounds strange. Sure I can have a look at log see if I can spot the problem.

**Comment by miladf2** [comment]
I think I made it work... I will copy my code here. I am trying to improve the code to be able to be able to score my dataset and possibly change the base from VGG to Inception or some other structures.

**Comment by balancap** [comment]
Great if you made it work! Would be happy to merge the fix :)
I tried to make the project as flexible as possible to be able to change to add architecture based on Inception or others.

**Comment by miladf2** [comment]
One other thing I noticed is if I retrain one of your pretrained models on another dataset of images of 1280*720 images, it will mess up the pretrained net. I think those ssd_vgg nets should have a resize in the beginning. They should have an IOU too so they dont detect objects on top of each other.

**Comment by balancap** [comment]
Is there other dataset different from Pascal VOC? If you need to have good performance on both sets, I think you need to train on both combines. I will try shortly to improve the performance of training script,  which not as good as the Caffe one, and allow training on multiple datasets combined.

During training, the preprocessing methods perform random crop on images and resize everything to 300 x 300 at the end. So the input size should not have an effect on training results (Pascal VOC dataset for instance have images of different sizes). For the IOU, it depends on the `threshold` you give to the NMS algorithm : the lower it is, the more it will tends to merge overlapping boxes.

**Comment by zsz02** [comment]
@miladf2 , I want train on 4GPU ,  but it do not work , can you  show me how to set options??  thank you !!

**Comment by vinayakarannil** [comment]
can i retrain this model for my own dataset?? for example if i need to detect traffic signals from images?

**Comment by skyloveblue** [comment]
I‘m trying to train on 4 GPUs, and I changed one code in train_ssd.py: trainconfig = tf.ConfigProto(log_device_placement=False,allow_soft_placement=True, gpu_options=gpu_options) and by using num_clones=8, followed @miladf2 But it dosen't work. Could you show me how to solve this problem? @miladf2 Thank you.
